### ISSUE TITLE
[test-diag] unit test linker error on OS X (10.13.4)

### ISSUE THREAD

**Issue by rdewaele** [openthread/openthread#2732]
## Platform Information

```
  Configuration Summary
  ---------------------
  Package                                   : OPENTHREAD
  Version                                   : 20170716-00602-g52644515-dirty
  Interface                                 : 0.0.5
  Build system                              : x86_64-apple-darwin17.5.0
  Host system                               : x86_64-apple-darwin17.5.0
  Host architecture                         : x86_64
  Host OS                                   : darwin17.5.0
  Cross compiling                           : no
  Build shared libraries                    : yes
  Build static libraries                    : yes
  Build debug libraries                     : no
  Build optimized libraries                 : yes
  Build coverage libraries                  : no
  Build coverage reports                    : no
  Address sanitizer support                 : no
  Lcov                                      : -
  Genhtml                                   : -
  Build tests                               : yes
  Build fuzz targets                        : no
  Build tools                               : yes
  OpenThread tests                          :
  Prefix                                    : /
  Documentation support                     : no
  Doxygen                                   : -
  GraphViz dot                              : -
  C Preprocessor                            : gcc -E
  C Compiler                                : gcc
  C++ Preprocessor                          : g++ -E
  C++ Compiler                              : g++
  Assembler Compiler                        : gcc
  Archiver                                  : ar
  Archive Indexer                           : ranlib
  Symbol Stripper                           : strip
  Object Copier                             :
  C Preprocessor flags                      : -O1 -g -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='"openthread-core-posix-config.h"' -I/private/tmp/openthread/examples/../examples/platforms/posix/   -I${abs_top_srcdir}/third_party/mbedtls -I${abs_top_srcdir}/third_party/mbedtls/repo.patched/include -I${abs_top_srcdir}/third_party/mbedtls/repo.patched/include/mbedtls -DMBEDTLS_CONFIG_FILE=\"mbedtls-config.h\"
  C Compile flags                           :  -O1 -g -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='"openthread-core-posix-config.h"' -I/private/tmp/openthread/examples/../examples/platforms/posix/   -Wall -Wextra -Wshadow -Werror -std=c99 -pedantic-errors -Wconversion -D_BSD_SOURCE=1 -D_DEFAULT_SOURCE=1
  C++ Compile flags                         :  -O1 -g -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='"openthread-core-posix-config.h"' -I/private/tmp/openthread/examples/../examples/platforms/posix/   -Wall -Wextra -Wshadow -Werror -std=gnu++98 -Wno-c++14-compat -Wconversion -D_BSD_SOURCE=1 -D_DEFAULT_SOURCE=1
  Assembler flags                           :  -O1 -g -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='"openthread-core-posix-config.h"' -I/private/tmp/openthread/examples/../examples/platforms/posix/
  Link flags                                : -O1 -g -DOPENTHREAD_PROJECT_CORE_CONFIG_FILE='"openthread-core-posix-config.h"' -I/private/tmp/openthread/examples/../examples/platforms/posix/
  Link libraries                            :
  Link maps                                 : no
  Pretty                                    : ${abs_top_srcdir}/script/clang-format.sh
  Pretty args                               : -style=file -i
  Pretty check                              : ${abs_top_srcdir}/script/clang-format-check.sh
  Pretty check args                         : -
  OpenThread CLI support                    : yes
  OpenThread CLI-MTD support                : yes
  OpenThread CLI-FTD support                : yes
  OpenThread NCP support                    : yes
  OpenThread NCP-MTD support                : yes
  OpenThread NCP-FTD support                : yes
  OpenThread NCP-BUS Configuration          : uart
  OpenThread NCP Vendor Hook Source         : none
  OpenThread NCP Spinel Encrypter           : none
  OpenThread Multiple Instances support     : no
  OpenThread MTD Network Diagnostic support : yes
  OpenThread builtin mbedtls support        : yes
  OpenThread TMF Proxy support              : yes
  OpenThread Commissioner support           : yes
  OpenThread Joiner support                 : yes
  OpenThread DTLS support                   : yes
  OpenThread Jam Detection support          : yes
  OpenThread Channel Monitor support        : no
  OpenThread Channel Manager support        : no
  OpenThread MAC Filter support             : yes
  OpenThread Diagnostics support            : yes
  OpenThread Child Supervision support      : yes
  OpenThread Legacy network support         : yes
  OpenThread Certification log support      : yes
  OpenThread DHCPv6 Server support          : yes
  OpenThread DHCPv6 Client support          : yes
  OpenThread DNS Client support             : yes
  OpenThread Application CoAP support       : yes
  OpenThread Raw Link-Layer support         : yes
  OpenThread Border Router support          : yes
  OpenThread Service support                : yes
  OpenThread examples                       : posix
```

```
❯ gcc -v
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.1.0 (clang-902.0.39.1)
Target: x86_64-apple-darwin17.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

```
❯ g++ -v
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.1.0 (clang-902.0.39.1)
Target: x86_64-apple-darwin17.5.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

## Steps to produce the error

```
❯ git rev-parse --short HEAD
52644515
❯ make -f examples/Makefile-posix
<snip>
❯ cd build/x86_64-apple-darwin
build/x86_64-apple-darwin❯ make check
<snip>
make[4]: `test-toolchain' is up to date.
  CXX      test_diag.o
  CXXLD    test-diag
Undefined symbols for architecture x86_64:
  "_gPlatformPseudoResetWasRequested", referenced from:
      _PlatformInit in libopenthread-posix.a(libopenthread_posix_a-platform.o)
      _PlatformPseudoResetWasRequested in libopenthread-posix.a(libopenthread_posix_a-platform.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[4]: *** [test-diag] Error 1
<snip>
```

## Symbol living in the common section

Inspection of the archive shows that the missing symbol _is_ present in the misc.o object file in the _common_ section. It is also undefined in the platform.o object file. This corresponds with the definition of the variable in misc.c file and the extern declaration in the platform.c file.

```
❯ nm examples/platforms/posix/libopenthread-posix.a | grep -B14 _gPlatformPseudoResetWasRequested
<snip>
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o):
                 U _PlatformDeinit
                 U _alarm
                 U _execvp
                 U _exit
                 U _gArguments
                 U _gArgumentsCount
0000000000001120 b _gPlatMcuPowerState
0000000000000001 C _gPlatformPseudoResetWasRequested
--
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-platform.o):
00000000000002c4 D _NODE_ID
0000000000000140 T _PlatformDeinit
0000000000000000 T _PlatformInit
0000000000000150 T _PlatformProcessDrivers
0000000000000130 T _PlatformPseudoResetWasRequested
00000000000002c8 D _WELLKNOWN_NODE_ID
                 U ___error
                 U ___stderrp
                 U _basename
                 U _exit
                 U _fprintf
00000000000011e0 S _gArguments
00000000000011d8 S _gArgumentsCount
                 U _gPlatformPseudoResetWasRequested
```

(`nm` documentation quotes from https://sourceware.org/binutils/docs/binutils/nm.html)

> The symbol is common. Common symbols are uninitialized data. When linking, multiple common symbols may appear with the same name. If the symbol is defined anywhere, the common symbols are treated as undefined references. For more details on common symbols, see the discussion of –warn-common in Linker options in The GNU linker.

This prompted me to *add an initialization* to the definition in `misc.c`.

```
 48 bool                       gPlatformPseudoResetWasRequested = false;
```

Which fixes the issue. Upon further inspection we can also see that the variable no longer lives in the _common_ section, but in a section marked 'S':
> The symbol is in an uninitialized data section for small objects.

```
❯ nm examples/platforms/posix/libopenthread-posix.a | grep -B14 _gPlatformPseudoResetWasRequested
<snip>
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o):
                 U _PlatformDeinit
                 U _alarm
                 U _execvp
                 U _exit
                 U _gArguments
                 U _gArgumentsCount
0000000000001120 b _gPlatMcuPowerState
000000000000111e S _gPlatformPseudoResetWasRequested
--
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-platform.o):
00000000000002c4 D _NODE_ID
0000000000000140 T _PlatformDeinit
0000000000000000 T _PlatformInit
0000000000000150 T _PlatformProcessDrivers
0000000000000130 T _PlatformPseudoResetWasRequested
00000000000002c8 D _WELLKNOWN_NODE_ID
                 U ___error
                 U ___stderrp
                 U _basename
                 U _exit
                 U _fprintf
00000000000011e0 S _gArguments
00000000000011d8 S _gArgumentsCount
                 U _gPlatformPseudoResetWasRequested
```

## But the examples link properly!?

The problem only occurs when compiling this specific unit test, but not with the example applications. This seemed quite odd to me because the linker error above only relates to objects within a single archive, so I investigated further.

By running `make V=1` and `make V=1 check` I isolated the specific libtool commands used to link a cli app and the failing unit test respectively. I added diagnostic flags to the linker and ran the commands again.

- [vanilla_notfound.log](https://github.com/openthread/openthread/files/2029111/vanilla_notfound.log) shows that the linker fails to load the misc.o object file, which of course results in the symbol not being found.

- [vanilla_all_load.log](https://github.com/openthread/openthread/files/2029164/vanilla_all_load.log) shows that forcing the linker to load all objects results in a successful link.

```
<snip>
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
<snip>
```

- [initialized_found.log](https://github.com/openthread/openthread/files/2029167/initialized_found.log) shows that with the modification described above (initialize the global variable) directly causes the linker to load the misc.o object file.

```
<snip>
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
_gPlatformPseudoResetWasRequested forced load of ../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
<snip>
```

- [cli_link.log](https://github.com/openthread/openthread/files/2029188/cli_link.log) shows that in the case of the cli application, another piece of code causes the misc.o object file to be loaded.

```
<snip>
../../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
_otPlatReset forced load of ../../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
<snip>
```

- [platreset_found.log](https://github.com/openthread/openthread/files/2029215/platreset_found.log) shows that hacking this behaviour into `test_diag.cpp` also fixes the linker issue. With the following bit added to the end of the file (and _without_ initializing the global variable):

```
void unused() {
    otPlatReset(NULL);
}
```

```
<snip>
../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
_otPlatReset forced load of ../../examples/platforms/posix/libopenthread-posix.a(libopenthread_posix_a-misc.o)
<snip>
```

## Summary

It seems that there are multiple ways in which this problem can be solved. I am not sure which is the best way to go, but I will share my personal opinions below:

I think that tricking the linker to include some object files (as accidentally happens with the application linking process) is not a solution to the core problem.

Reading the discussion for the `--warn-common` linker option on https://sourceware.org/binutils/docs/ld/Options.html#Options leads me to believe that this is not a feature that should be relied upon in modern, portable C code. To share a variable across different compilation units, a declaration in a designated header in combination with a single definition seems to be a strictly better alternative to me.
While both `int foo;` and `int foo = 42;` are variable definitions, compilers may still make a distinction by way of this (very common) extension (i.e. it technically violates the single definition rule) to standard C (excerpt from `man clang`):

```
-fcommon, -fno-common
              This flag specifies that variables without initializers get common linkage.  It can be  dis-
              abled with -fno-common.
```

Using this compiler switch also solves the linker error:

```
❯ CFLAGS=-fno-common CXXFLAGS=-fno-common CPPFLAGS=-fno-common make -j8 -f examples/Makefile-posix
<snip>
❯ nm build/x86_64-apple-darwin/examples/platforms/posix/libopenthread-posix.a | grep _gPlatformPseudoResetWasRequested
000000000000112c S _gPlatformPseudoResetWasRequested
                 U _gPlatformPseudoResetWasRequested
❯ cd build/x86_64-apple-darwin
build/x86_64-apple-darwin❯ make check
<snip>
  CXX      test_diag.o
  CXXLD    test-diag
  CXX      test_ncp_buffer.o
  CXXLD    test-ncp-buffer
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: test-aes
<snip>
```

So I think that if global variables _must_ be used, direct initialization will make sure they will not have common linkage, while also making sure that they are always properly initialized. Adding `-fno-common` or `-Wl,-load_all` (or equivalent) moves the fix to the build system and is compatible with the former.

P.S.: While I have found a number of solutions to make the problem go away, I am still not sure _why_ the linker refuses to load the misc.o object file all by itself and am really curious if this is normal behavior. :)

**Comment by jwhui** [comment]
@rdewaele, thank you for raising this issue and detailed description!  If I read this correctly, your suggestion is to initialize globals at declaration.  If so, I agree and please submit a PR with the fix.

**Comment by rdewaele** [comment]
@jwhui Indeed, initialising the globals fixed the issue in my testing. The alternative/complementary fix is to add some flags to the build system (compiler `-fno-common` and/or linker `-Wl,-load_all`) but these are more heavy handed I feel, and may have side effects.

Would you like me to scan the whole project for uninitialised globals, or change the one I experienced this problem with.

**Comment by jwhui** [comment]
@rdewaele, I agree with not relying on compiler flags to address this. If you are willing, it would be great if you addressed all uninitialized globals.

**Comment by rdewaele** [comment]
@jwhui Just a heads up, it's likely I'll only have time in the weekends for a bit to work on this, so it may take a week or two before I get around to wrap this up.

**Comment by jwhui** [comment]
@rdewaele, any update on your progress?

**Comment by rdewaele** [comment]
@jwhui I took some time off this and next weekend so I'll try and look into this either this or next Friday. Our project is currently focusing on parts different than OpenThread so that's why I didn't find the time yet.

**Comment by bukepo** [comment]
Close this issue for #3049?

**Comment by jwhui** [comment]
Resolved by #3049.
